### PR TITLE
Update IdPs/SPs from SAML metadata

### PR DESCRIPTION
**Pull request by James-REANNZ**
Some IdPs/SPs are added directly to the SAML service. The reporting service sources it's information from the federation registry, meaning it doesn't see those entities. This change pulls metadata from the SAML service to augment the data from the federation registry.

### CODE REVIEW

File: .rubocop.yml [hunk 1]
```diff
@@ -30,3 +30,6 @@ Rails/InverseOf:
 Lint/MissingSuper:
   Enabled: false
 
+Metrics/ClassLength:
+  Exclude:
+    - app/jobs/update_from_saml_service.rb
```
Comment (vladimir-mencl-eresearch) [inline .rubocop.yml]: In the AAF code base, I've generally seen per-class exclusions done inline, for from `.rubocop.yml`

File: Gemfile.lock [hunk 2]
```diff
@@ -370,17 +370,20 @@ GEM
...
 
 PLATFORMS
   x86_64-darwin-20
+  x86_64-darwin-21
   x86_64-linux
 
 DEPENDENCIES
   aaf-gumboot
-  aaf-lipstick (>= 4.5.1)
+  aaf-lipstick (>= 4.5.1)!
```
Comment (vladimir-mencl-eresearch) [inline Gemfile.lock]: Why the exclamation mark at the end of the line?

**Comment by vladimir-mencl-eresearch** [comment]
Hi @James-REANNZ ,

Thanks for the PR and sorry about the long delay in getting back to you.  School holidays and then sick children got in the way.  (Plus when I saw you were already working out the thoughts for a replacement, I thought this might be low priority anyway....)

But here are my thoughts on the PR.

(1) I like how you approached this with parsing the metadata from the SAML service (and found code to reuse for dealing with the metadata parsing).

(2) I would like to see a clear position on what remains as the role of the original sync from the Federation Registry.  I see the `clean` method is modified to only purge attribute definitions that have not been seen by the sync.  But it appears it would still attempt to synchronise (copy in) everything it used to sync.

All IdPs and SPs seen in the FR would also be seen in SAML Service.  So it might be right to remove them from the scope of the sync.

A tricky bit is that Organizations would only be seen in the SAML Service metadata if they have at least one IdP or SP - while FR also exposes Organizations that are "empty shells".  But the only benefit of having these Organizations visible in the Reporting Service is for the report on Federation Growth (number of registrations) - which I do not see as this crucial.

Perhaps this is more a question for the new Proposal for Reporting Service replacement (whether to sync empty Organizations).

But my main point here is to see a clear position on how the new sync from SAML Service interacts with the existing sync from FR.

But, if the new proposal moves forward, all of this may be just academic discussion.

I have a bit more feedback on individual aspects and I'll try attaching that to specific lines in the code.

But this is the main summary: like the approach, but want to see the clear position on how the two mechanisms would interact.

And thanks for getting it into the PR!

Cheers,
Vlad

File: app/jobs/update_from_saml_service.rb [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+    OpenSSL::X509::Certificate.new(cert) if cert
+  end
+
+  def configuration
+    Rails.application.config.reporting_service.saml_service
+  end
+
+  def org_identifier(name)
+    # Generate a valid identifier for an organisation given the OrganisationName (domain)
+    digest = OpenSSL::Digest.new('SHA256').digest("tuakiri:subscriber:#{name}")
```
Comment (vladimir-mencl-eresearch) [inline app/jobs/update_from_saml_service.rb]: I'm a little concerned about having two `org_identifier` methods that provide the same function but implement it differently - potential source for confusion.
https://github.com/REANNZ/reporting-service/blob/master/app/jobs/concerns/query_federation_registry.rb#L13

File: app/jobs/update_from_saml_service.rb [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+
+    org.update!(org_attrs)
+
+    org
+  end
+
+  def process_idp(node, entity_id, org, reg_date)
+    idp = IdentityProvider.find_or_initialize_by(entity_id: entity_id)
+
+    name_node = xpath_at(node, './md:Extensions/mdui:UIInfo/mdui:DisplayName')
```
Comment (vladimir-mencl-eresearch) [inline app/jobs/update_from_saml_service.rb]: Watch out - there might be multiple instances of `mdui:DisplayName`, distinguished by the `xml:lang` attribute.  You'd want `xml:lang="en"`.  In Tuakiri, you'd likely only see the `en` versions, but eduGAIN-imported metadata might have more languages.

File: app/jobs/update_from_saml_service.rb [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+    process_idp_attributes(idp_attributes, node)
+
+    @identity_providers.append(idp)
+    activate_object(idp, reg_date)
+  end
+
+  def process_sp(node, entity_id, org, reg_date)
+    sp = ServiceProvider.find_or_initialize_by(entity_id: entity_id)
+
+    name_node = xpath_at(node, './md:Extensions/mdui:UIInfo/mdui:DisplayName')
```
Comment (vladimir-mencl-eresearch) [inline app/jobs/update_from_saml_service.rb]: Same comment - should expect multiple instances and pick the `en` version.

File: app/jobs/update_from_saml_service.rb [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+    idp_node = xpath_at(node, './md:IDPSSODescriptor')
+    process_idp(idp_node, entity_id, org, reg_date) if idp_node
+
+    sp_node = xpath_at(node, './md:SPSSODescriptor')
+    process_sp(sp_node, entity_id, org, reg_date) if sp_node
+  end
+
+  def process_org(node)
+    org_id_node = xpath_at(node, './md:OrganizationName')
+    org_name_node = xpath_at(node, './md:OrganizationDisplayName')
```
Comment (vladimir-mencl-eresearch) [inline app/jobs/update_from_saml_service.rb]: Same comment - should expect multiple instances and pick the `en` version.

File: config/reporting_service.yml.dist [hunk 7]
```diff
@@ -15,6 +15,9 @@ federationregistry:
 rapidconnect:
   host: rapid.test.aaf.edu.au
   secret: 'This is the shared secret used for authenticating to the Rapid export API'
+saml_service:
+  metadata_url: https://saml.test.url/mdq
+  # metadata_cert: "The Certificate"
```
Comment (vladimir-mencl-eresearch) [inline config/reporting_service.yml.dist]: It wasn't clear to me right away whether this was meant to be a path to the certificate or the certificate embedded directly.  Probably obvious (and just not to me), but maybe the sample value could make it even clearer?

File: db/migrate/20220413213421_add_domain_to_organisation.rb [hunk 8]
```diff
@@ -0,0 +1,6 @@
+class AddDomainToOrganisation < ActiveRecord::Migration[6.0]
```
Comment (vladimir-mencl-eresearch) [inline db/migrate/20220413213421_add_domain_to_organisation.rb]: I have seen AAF committing the magic string:
`# frozen_string_literal: true`
as the first line of all Ruby code, including migrations - I expect Rubocop will complain if it isn't there...

**Review by vladimir-mencl-eresearch — CHANGES_REQUESTED** [review]

**Comment by vladimir-mencl-eresearch** [comment]
Done with the review. Happy to discuss offline (Slack or Zoom) - though not sure how much this is worth pursuing - also depends on how the Reporting Service Replacement proposal progresses.

File: db/migrate/20220413213421_add_domain_to_organisation.rb (hunk 8 above)
Comment (vladimir-mencl-eresearch) [reply db/migrate/20220413213421_add_domain_to_organisation.rb]: No longer required by Rubocop.

File: Gemfile.lock (hunk 2 above)
Comment (vladimir-mencl-eresearch) [reply Gemfile.lock]: Got it in master in the meantime as well - I assume newer bundler syntax.

File: .rubocop.yml (hunk 1 above)
Comment (vladimir-mencl-eresearch) [reply .rubocop.yml]: The `.rubocop.yml` file needs rebasing anyway, I'll change this in the rebase.

File: config/reporting_service.yml.dist (hunk 7 above)
Comment (vladimir-mencl-eresearch) [reply config/reporting_service.yml.dist]: As  configuration is now via env vars, which is not exactly suitable for passing full certificates, I'll change the setting to use path to certificate instead.

File: app/jobs/update_from_saml_service.rb (hunk 4 above)
Comment (vladimir-mencl-eresearch) [reply app/jobs/update_from_saml_service.rb]: Sorted.

File: app/jobs/update_from_saml_service.rb (hunk 5 above)
Comment (vladimir-mencl-eresearch) [reply app/jobs/update_from_saml_service.rb]: Sorted

File: app/jobs/update_from_saml_service.rb (hunk 6 above)
Comment (vladimir-mencl-eresearch) [reply app/jobs/update_from_saml_service.rb]: Sorted

File: app/jobs/update_from_saml_service.rb (hunk 3 above)
Comment (vladimir-mencl-eresearch) [reply app/jobs/update_from_saml_service.rb]: Renamed to avoid confusion.

**Review by vladimir-mencl-eresearch — APPROVED** [review]

File: app/jobs/update_from_saml_metadata.rb [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+    process_idp(idp_node, entity_id, org, reg_date) if idp_node
+
+    sp_node = xpath_at(node, './md:SPSSODescriptor')
+    process_sp(sp_node, entity_id, org, reg_date) if sp_node
+  end
+
+  # rubocop:disable Metrics/MethodLength, Metrics/AbcSize
+  def process_org(node)
+    org_domain_node = xpath_at(node, './md:OrganizationName')
+    org_name_node = xpath_at(node, "./md:OrganizationDisplayName[@xml:lang='en']") ||
```
Comment (James-REANNZ) [inline app/jobs/update_from_saml_metadata.rb:49]: Do we want to handle the case where there isn't an english name, but there is a non-english name? Or is that not worth handling?

**Review by James-REANNZ — COMMENTED** [review]
Looks good to me (and not just because half of it is my work :P).

File: app/jobs/update_from_saml_metadata.rb (hunk 9 above)
Comment (vladimir-mencl-eresearch) [reply app/jobs/update_from_saml_metadata.rb:49]: Thanks, good point.  I thought of it as an edge-case with a question mark.  But if it ever gets used to ingest eduGAIN metadata, anything should be expected - including no name at all.

Pushed now - thanks for the suggestion.

**Pull request merged**